### PR TITLE
DAE support in SBML

### LINKED ISSUE #760: Python SBML import: add support for algebraic rules

### PR DESCRIPTION
**Pull request by FFroehlich**
fixes #760

### CODE REVIEW

**Comment by FFroehlich** [comment]
tbd: rename xdot symbols

File: python/sdist/amici/sbml_import.py [hunk 1]
```diff
@@ -1027,6 +1006,137 @@ def _process_rules(self) -> None:
...
+            is_involved_in_reaction = is_species and \
+                not smart_is_zero_matrix(self.stoichiometric_matrix[
+                    list(self.symbols[SymbolId.SPECIES].keys()).index(symbol),
+                    :
+                ])
+            if is_species and not is_boundary_condition and is_involved_in_reaction:
+                continue
+            free_variables.add(symbol)
+
+        assert len(free_variables) >= 1
```
Comment (dweindl) [inline python/sdist/amici/sbml_import.py:1049]: Is that supposed to be caught by SBML validation? If so, good. If not, a more informative message would be nice.

File: python/sdist/amici/sbml_import.py [hunk 2]
```diff
@@ -1027,6 +1006,137 @@ def _process_rules(self) -> None:
...
+            'value': formula
+        }
+        # remove the symbol from the original definition and add to
+        # algebraic symbols (if not already done)
+        for var in free_variables:
+            if var in self.symbols[SymbolId.FIXED_PARAMETER]:
+                raise SBMLException(
+                    'There are algebraic rules that specify the '
+                    f'value of {var}, which is also marked as '
+                    'fixed paramter. This is currently not supported!'
```
Comment (dweindl) [inline python/sdist/amici/sbml_import.py:1062]:
```suggestion
                    'fixed parameter. This is currently not supported! '
```

File: python/sdist/amici/sbml_import.py [hunk 3]
```diff
@@ -1027,6 +1006,137 @@ def _process_rules(self) -> None:
...
+            if var in self.symbols[SymbolId.ALGEBRAIC_STATE]:
+                continue
+            if var in self.compartments:
+                init = self.compartments[var]
+                symbol = {
+                    'name': str(var),
+                    'value': init,
+                }
+                symbol_id = 'compartment'
+                six = np.nan
```
Comment (dweindl) [inline python/sdist/amici/sbml_import.py:1076]: `six`? `symbol_idx`?

File: src/model_dae.cpp [hunk 4]
```diff
@@ -117,15 +159,44 @@ void Model_DAE::fdxdotdp(const realtype t, const const_N_Vector x,
 void Model_DAE::fM(realtype t, const_N_Vector x) {
     derived_state_.M_.zero();
     auto x_pos = computeX_pos(x);
```
Comment (dweindl) [inline src/model_dae.cpp:161]: Move x_pos to matlab branch?

**Review by dweindl — COMMENTED** [review]
Awesome! Thanks!

File: include/amici/model_dae.h [hunk 5]
```diff
@@ -38,17 +38,28 @@ class Model_DAE : public Model {
...
+              const int ndxdotdp_explicit=0, const int ndxdotdx_explicit=0,
+              const int w_recursion_depth=0)
         : Model(model_dimensions, simulation_parameters,
                 o2mode, idlist, z2event, pythonGenerated,
-                ndxdotdp_explicit) {
+                ndxdotdp_explicit, ndxdotdx_explicit, w_recursion_depth) {
         derived_state_.M_ = SUNMatrixWrapper(nx_solver, nx_solver);
+            int ndiff = 0;
+        for (auto& n : idlist)
+             ndiff += static_cast<int>(n);
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply include/amici/model_dae.h:57]: IDAS expects floats: 
>id (N Vector) state vector. A value of 1.0 indicates a differential variable, while
0.0 indicates an algebraic variable.

File: python/sdist/amici/sbml_import.py [hunk 6]
```diff
@@ -1027,6 +1006,137 @@ def _process_rules(self) -> None:
...
+            if sbml_var.isSetConstant() and sbml_var.getConstant():
+                continue
+            # and there must also not be a rate rule or assignment
+            # rule for it
+            if self.is_assignment_rule_target(sbml_var) or \
+                    self.is_rate_rule_target(sbml_var):
+                continue
+            # Furthermore, if the entity is a Species object, its value
+            # must not be determined by reactions, which means that it
+            # must either have the attribute boundaryCondition=“false”
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply python/sdist/amici/sbml_import.py:1034]: this is a direct quote from the SBML spec. Took me a while to figure out, but this is indeed correct.

[the remark this answers is not in the document]
Comment (dweindl) [reply python/sdist/amici/sbml_import.py:1034]: Okay. I read it three times. I had the feeling the code checks the opposite. But if tests pass, I guess this is good.

File: python/sdist/amici/import_utils.py [hunk 7]
```diff
@@ -19,6 +19,11 @@
 except ImportError:
     pysb = None
 
+
+class SBMLException(Exception):
+    pass
+
+
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply python/sdist/amici/import_utils.py:26]: https://github.com/AMICI-dev/AMICI/blob/55204d16eeda37aa8639c8a9994f426f82202695/python/sdist/amici/de_export.py#L947

File: python/sdist/amici/sbml_import.py [hunk 8]
```diff
@@ -28,17 +29,15 @@
...
-from .ode_export import (
-    ODEExporter, ODEModel, symbol_with_assumptions, _default_simplify
+from .de_export import (
+    DEExporter, DEModel, symbol_with_assumptions, _default_simplify,
+    smart_is_zero_matrix
 )
 
 
-class SBMLException(Exception):
-    pass
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply python/sdist/amici/sbml_import.py:39]: https://github.com/AMICI-dev/AMICI/blob/55204d16eeda37aa8639c8a9994f426f82202695/python/sdist/amici/de_export.py#L947

File: python/sdist/amici/import_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dweindl) [reply python/sdist/amici/import_utils.py:26]: Okay. That function should also be moved to sbml_import at some point.

File: python/sdist/amici/de_export.py [hunk 9]
```diff
@@ -718,7 +750,9 @@ def __init__(self, verbose: Optional[Union[bool, int]] = False,
             Whether to cache calls to the simplify method. Can e.g. decrease
             import times for models with events.
         """
-        self._states: List[State] = []
+        self._differentialstates: List[DifferentialState] = []
+        self._algebraicstates: List[AlgebraicState] = []
+        self._algebraicequations: List[AlgebraicEquation] = []
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply python/sdist/amici/de_export.py:755]: correct

**Review by dweindl — APPROVED** [review]

**Pull request merged**